### ISSUE TITLE
Bug/Question - Trait methods can (apparently) be overridden.

### ISSUE THREAD

**Issue by hotoiledgoblinsack** [Peekmo/atom-autocomplete-php#177]
Hello

Today I learned that *apparently*, a method that originates from a trait can be overridden by a method from the class using the trait. I was totally stumped when I learned this as this looks like function overloading in my eyes, which PHP does not support:

```php
trait A
{
    public function foo($param1) {}
}

class B
{
    use A;

    // Apparently completely valid.
    public function foo($param1) {}
}
```

In any case, I'm not sure if it's very good practice to use it, as you can alias the trait method to use a different name such as `fooPrivate`.

I would like to support this, but the problem is, for goto at least, I can't uniquely detemine the `declaringStructure` of the method. Currently it works like this:
  * Fetch `declaringClass` of method. This is the (base) class that has it.
  * Look at traits of class to see if any of them have the method. If so, the method originates from the trait.

With this however, the `declaringStructure` of `foo` should be `B` as it takes precedence over the method from the trait, but when fetching methods for `B`, it will always include `foo`, no matter if it's overridden there or not (as `getMethods` also includes trait methods). Without fixing this, alt-clicking a call to `->foo` will always navigate to the trait method instead of to the class method.

You could always regex your way through the class file to look for the overload, but that seems like a bad idea. Any idea's and should we even support this?

PS: If we support this, `isAbstract` checks in `Tools` also have to be removed for trait methods as non-abstract methods can also be overridden.

**Comment by Peekmo** [comment]
When I see all the implications, I think that "if" we support it, it will be after a first release.

For the moment, I think that it will be a ```wontfix``` as it is a really not common case.

**Comment by hotoiledgoblinsack** [comment]
Ok, seems like a good plan to me. Perhaps we should put this under "Things that will not be implemented" in the README, along with #101?

**Comment by hotoiledgoblinsack** [comment]
I'm closing this as it is won't fix and mentioned in the README for future reference.
